packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts

```diff
@@ -0,0 +1,88 @@
+import * as React from 'react';
+
+/**
+ * Checks if a React element has all of the specified props explicitly provided.
+ *
+ * @internal - Exported for testing purposes
+ */
+export function hasExplicitProps(element: React.ReactElement, props: string[]): boolean {
+  if (!element.props || typeof element.props !== 'object') {
+    return false;
+  }
+
+  for (const prop of props) {
+    if (!(prop in element.props)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/**
+ * Checks if a React element is a motion component created by createMotionComponent.
+ * Motion components have the function name "Atom".
+ *
+ * **Note:** This is a heuristic detection. Motion components may or may not support
+ * specific props like `delay` or `visible` depending on their implementation.
+ *
+ * @internal - Exported for testing purposes
+ */
+export function isMotionComponent(element: React.ReactElement): boolean {
+  return typeof element.type === 'function' && element.type.name === 'Atom';
```

Comment:
This is not reliable as `.name` will be changed in production build by a minifier